### PR TITLE
POC data entry

### PR DESCRIPTION
**Pull request by mestachs**
currently support : 
- [x] try to keep the "specific code" dumb and more declarative
- [x] read-write
- [x] dhis2 indicator "a like" (only if current orgunit data)
- [x] plugin based

Todo

- [x] completed dataset status
- [ ] better hesabu integration (think package, activity, states/formulas ...)
- [x] nicer screen/better integration with invoices and contracts

See the "specific" code for a data entry : https://github.com/BLSQ/blsq-report-specific-sandox/pull/1
you have a DataEntries that receives an orgunit (with contracts) and period and gives info about the available data entries
and can provide dedicated Form for data entry, reusing provided components
```js
  <Dhis2Input dataElement="lXbpVIzzwef.HllvX50cXC0" />
  <Dhis2Input dataElement="VBIYcOJSmpU" />

  <Dhis2Formula formula="ROUND( ABS(SAFE_DIV( #{lXbpVIzzwef.HllvX50cXC0} - #{VBIYcOJSmpU.HllvX50cXC0},#{lXbpVIzzwef.HllvX50cXC0} + #{VBIYcOJSmpU.HllvX50cXC0}) * 100),2)" />
```

you can end-up to something like this

image

image

at some point I hope to have them in a more central config, so data iaso/hesabu can "know" about this.

### CODE REVIEW

File: src/components/dataentry/DataEntrySelectionPage.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      );
+
+      const dv = await api.get("/dataValueSets", {
+        dataSet: dataEntry.dataSetId,
+        orgUnit: activeContract.orgUnit.id,
+        period: period,
+      });
+      const rawValues = dv.dataValues || [];
+
+      const defaultCoc = "HllvX50cXC0";
```
Comment (vanakenm) [inline src/components/dataentry/DataEntrySelectionPage.js:54]: Could this be in the "custom part"? Or some kind of config?

**Comment by mestachs** [comment]
for more hesabu integration...

1. webasm with go engine
2. big eval with something like this : https://gist.github.com/mestachs/c848348dbe72ccb17c8896991ed0186e

but still need something that "looks like" a simulation json (activity item, total items with their instantiated formulas, but not the "inputs" only the incentives)
not all packages can work (eg the one with ponderation with zone, ...)
how to know where to stop ?

File: src/components/dataentry/DataEntrySelectionPage.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  const [formData, setFormData] = useState(undefined);
+  const [error, setError] = useState(undefined);
+  useEffect(() => {
+    const loadData = async () => {
+      const contractService = PluginRegistry.extension("contracts.service");
+
+      const period = match.params.period;
+      const contracts = await contractService.fetchContracts(match.params.orgUnitId);
+      const activeContract = contracts.mainContracts.contracts.filter((c) => c.matchPeriod(period))[0];
+      if (activeContract == undefined) {
```
Comment (pjaspers) [inline src/components/dataentry/DataEntrySelectionPage.js:24]: Don't we need something like `typeof activeContract === undefined`?

File: src/components/dataentry/DataEntrySelectionPage.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        setError({
+          message: match.params.orgUnitId + " has no contract for that period : " + period,
```
Comment (pjaspers) [inline src/components/dataentry/DataEntrySelectionPage.js:26]: Do we use some kind of i18n here?

File: src/components/dataentry/Dhis2Formula.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import React, { useEffect, useState, useContext } from "react";
+import { TextField, Tooltip } from "@material-ui/core";
+import FormDataContext from "./FormDataContext";
+
+const parseDependencies = (expression) => {
+  const components = expression
```
Comment (pjaspers) [inline src/components/dataentry/Dhis2Formula.js:6]: I could use a comment here with an example

File: src/components/dataentry/Dhis2Formula.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      let expression = formula;
+      dependencies.forEach((dep) => {
+        const val = formDataContext.getValue(dep);
+        const valExpression = val !== undefined ? val.value : 0;
+        expression = expression.split("#{" + dep + "}").join(" " + valExpression + " ");
+      });
+
+      setExpression(expression);
+      try {
+        const evaluatedExpression = new Function(
```
Comment (pjaspers) [inline src/components/dataentry/Dhis2Formula.js:58]: Veru cool stuff.

File: src/components/dataentry/Dhis2Formula.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    return <></>;
+  }
+
+  return (
+    <div>
+      <Tooltip
+        title={
+          <div>
+
+              {formula} <br/><br/>
```
Comment (pjaspers) [inline src/components/dataentry/Dhis2Formula.js:84]: Maybe better to put in an <ul>

File: src/components/dataentry/useDebounce.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+const useDebounce = (obj, wait = 1250) => {
+  const [state, setState] = useState(obj);
+
+  const setDebouncedState = (val) => {
+    debounce(val);
+  };
+
+  const debounce = useCallback(
+    _.debounce((prop) => {
```
Comment (pjaspers) [inline src/components/dataentry/useDebounce.js:12]: TIL _.debounce 👍

**Review by pjaspers — APPROVED** [review]

File: src/components/dataentry/Dhis2Formula.js (hunk 5 above)
Comment (mestachs) [reply src/components/dataentry/Dhis2Formula.js:58]: this is way cooler : hesabu engine without dependency/parsing (need to replace a by this.a()) https://gist.github.com/mestachs/c848348dbe72ccb17c8896991ed0186e

**Pull request merged**